src/app/loadout/plug-drawer/PlugDrawer.tsx

```diff
@@ -112,50 +129,76 @@ export default function PlugDrawer({
+          const selected = oldState[plugSetHash] || [];
+          const firstIndex = selected.findIndex((s) => s.hash === plug.hash);
+          if (firstIndex !== -1) {
+            const newSelected = [...selected];
+            newSelected.splice(firstIndex, 1);
+            const newState = { ...oldState };
+            newState[plugSetHash] = newSelected;
+            return newState;
+          }
+        }
+        return oldState;
+      });
+    },
+    [setSelected]
+  );
+
   const onSubmit = (e: React.FormEvent | KeyboardEvent, onClose: () => void) => {
     e.preventDefault();
-    onAccept(selected);
+    onAccept(_.compact(Object.values(selected).flat()));
     onClose();
   };
 
-  const queryFilteredPlugs = useMemo(() => {
+  const queryFilteredPlugSets = useMemo(() => {
     const regexp = startWordRegexp(query, language);
-    return query.length
-      ? plugs.filter(
-          (plug) =>
-            regexp.test(plug.displayProperties.name) ||
-            regexp.test(plug.displayProperties.description) ||
-            regexp.test(plug.itemTypeDisplayName) ||
-            plug.perks.some((perk) => {
-              const perkDef = defs.SandboxPerk.get(perk.perkHash);
-              return (
-                perkDef &&
-                (regexp.test(perkDef.displayProperties.name) ||
-                  regexp.test(perkDef.displayProperties.description) ||
-                  regexp.test(perk.requirementDisplayString))
-              );
-            })
-        )
-      : plugs;
-  }, [query, plugs, defs.SandboxPerk, language]);
-
-  const groupedPlugs = Object.values(
-    _.groupBy(queryFilteredPlugs, (plugItem) => plugItem.plug.plugCategoryHash)
-  );
+    const rtn: PlugsWithMaxSelectable[] = [];
+
+    const searchFilter = (plug: PluggableInventoryItemDefinition) =>
+      regexp.test(plug.displayProperties.name) ||
+      regexp.test(plug.displayProperties.description) ||
+      regexp.test(plug.itemTypeDisplayName) ||
+      plug.perks.some((perk) => {
+        const perkDef = defs.SandboxPerk.get(perk.perkHash);
+        return (
+          perkDef &&
+          (regexp.test(perkDef.displayProperties.name) ||
+            regexp.test(perkDef.displayProperties.description) ||
+            regexp.test(perk.requirementDisplayString))
+        );
+      });
+
+    for (const { plugs, maxSelectable, plugSetHash } of plugsWithMaxSelectableSets) {
+      rtn.push({
+        plugSetHash,
+        maxSelectable,
+        plugs: query.length ? plugs.filter(searchFilter) : plugs,
+      });
+    }
+
+    return rtn;
+  }, [query, plugsWithMaxSelectableSets, defs.SandboxPerk, language]);
 
   if (sortPlugGroups) {
-    groupedPlugs.sort(sortPlugGroups);
+    queryFilteredPlugSets.sort(sortPlugGroups);
   }
 
   const autoFocus = !isPhonePortrait && !isiOSBrowser();
 
   const footer = ({ onClose }: { onClose(): void }) => (
     <Footer
-      selected={selected}
+      selected={_.compact(Object.values(selected).flat())}
```

Comment:
Wanna pull this into a variable so you can reuse it on line 243?

Reply:
Can do